java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -196,8 +194,30 @@ public void visitContinueStatement(ContinueStatementTree tree) {
   @Override
   public void visitMethodInvocation(MethodInvocationTree tree) {
     MethodInvocationTreeImpl mit = (MethodInvocationTreeImpl) tree;
-    if(mit.isTypeSet()) {
-      // nothing special to deduce once we have set type of expression
+    if(mit.isTypeSet() && mit.symbol().isMethodSymbol()) {
+      JavaSymbol.MethodJavaSymbol methodSymbol = (JavaSymbol.MethodJavaSymbol) mit.symbol();
+      JavaType methodReturnedType = (JavaType) mit.symbolType();
+      TypeSubstitution typeSubstitution = new TypeSubstitution();
```

Comment:
Th computation of the type substitution should be placed in a dedicated method, in order to allow unit testing.